File: nvmolkit/filterCatalog.cpp
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

#include <GraphMol/ROMol.h>

#include <boost/noncopyable.hpp>
#include <boost/python.hpp>
#include <boost/python/stl_iterator.hpp>
#include <cstdint>
#include <string>
#include <unordered_map>
#include <vector>

#include "src/substruct/filter_catalog.h"

namespace {

using namespace boost::python;

class ScopedGilRelease {
 public:
  ScopedGilRelease() : state_(PyEval_SaveThread()) {}
  ScopedGilRelease(const ScopedGilRelease&)            = delete;
  ScopedGilRelease& operator=(const ScopedGilRelease&) = delete;
  ~ScopedGilRelease() { PyEval_RestoreThread(state_); }

 private:
  PyThreadState* state_;
};

nvMolKit::FilterCatalogProperties toProperties(const dict& metadata) {
  nvMolKit::FilterCatalogProperties result;
  const list                        keys = metadata.keys();
  for (Py_ssize_t index = 0; index < len(keys); ++index) {
    const std::string key = extract<std::string>(keys[index]);
    result.emplace(key, extract<std::string>(metadata[key]));
  }
  return result;
}

std::vector<const RDKit::ROMol*> collectMolecules(const object& molecules, std::vector<object>& owners) {
  std::vector<const RDKit::ROMol*> pointers;
  stl_input_iterator<object>       iterator(molecules), end;
  for (; iterator != end; ++iterator) {
    owners.push_back(*iterator);
    pointers.push_back(extract<const RDKit::ROMol*>(owners.back()));
  }
  return pointers;
}

unsigned int addQuery(nvMolKit::FilterCatalog& catalog,
                      const RDKit::ROMol&      query,
                      const std::string&       description,
                      unsigned int             triggerCount,
                      const dict&              metadata) {
  auto                   properties = toProperties(metadata);
  const ScopedGilRelease release;
  return catalog.addEntry(query, description, triggerCount, std::move(properties));
}

unsigned int addSmarts(nvMolKit::FilterCatalog& catalog,
                       const std::string&       smarts,
                       const std::string&       description,
                       unsigned int             triggerCount,
                       const dict&              metadata) {
  auto                   properties = toProperties(metadata);
  const ScopedGilRelease release;
  return catalog.addSmarts(smarts, description, triggerCount, std::move(properties));
}

std::size_t addPreset(nvMolKit::FilterCatalog& catalog, std::uint32_t preset) {
  const ScopedGilRelease release;
  return catalog.addPreset(static_cast<nvMolKit::FilterCatalogPreset>(preset));
}

void finalize(nvMolKit::FilterCatalog& catalog) {
  const ScopedGilRelease release;
  catalog.finalize();
}

list hasMatch(nvMolKit::FilterCatalog& catalog, const object& molecules) {
  std::vector<object>       owners;
  const auto                pointers = collectMolecules(molecules, owners);
  std::vector<std::uint8_t> matches;
  {
    const ScopedGilRelease release;
    matches = catalog.hasMatch(pointers);
  }
  list result;
  for (const std::uint8_t match : matches) {
    result.append(match != 0);
  }
  return result;
}

list getFirstMatch(nvMolKit::FilterCatalog& catalog, const object& molecules) {
  std::vector<object> owners;
  const auto          pointers = collectMolecules(molecules, owners);
  std::vector<int>    matches;
  {
    const ScopedGilRelease release;
    matches = catalog.getFirstMatch(pointers);
  }
  list result;
  for (const int match : matches) {
    if (match < 0) {
      result.append(object());
    } else {
      result.append(match);
    }
  }
  return result;
}

list getMatches(nvMolKit::FilterCatalog& catalog, const object& molecules) {
  std::vector<object>                    owners;
  const auto                             pointers = collectMolecules(molecules, owners);
  std::vector<std::vector<unsigned int>> matches;
  {
    const ScopedGilRelease release;
    matches = catalog.getMatches(pointers);
  }
  list result;
  for (const auto& moleculeMatches : matches) {
    list ids;
    for (const unsigned int id : moleculeMatches) {
      ids.append(id);
    }
    result.append(ids);
  }
  return result;
}

dict getEntry(const nvMolKit::FilterCatalog& catalog, unsigned int id) {
  const auto entry = catalog.getEntry(id);
  dict       metadata;
  for (const auto& [key, value] : entry.properties) {
    metadata[key] = value;
  }
  dict result;
  result["id"]           = entry.id;
  result["description"]  = entry.description;
  result["smarts"]       = entry.smarts;
  result["triggerCount"] = entry.triggerCount;
  result["metadata"]     = metadata;
  return result;
}

}  // namespace

BOOST_PYTHON_MODULE(_filterCatalog) {
  import("rdkit.Chem.rdchem");
  import("nvmolkit._substructure");

  class_<nvMolKit::FilterCatalog, boost::noncopyable>(
    "FilterCatalog",
    init<nvMolKit::SubstructSearchConfig>((arg("config") = nvMolKit::SubstructSearchConfig())))
    .def("addPreset", &addPreset)
    .def("addQuery",
         &addQuery,
         (arg("query"), arg("description") = "", arg("triggerCount") = 1, arg("metadata") = dict()))
    .def("addSmarts",
         &addSmarts,
         (arg("smarts"), arg("description") = "", arg("triggerCount") = 1, arg("metadata") = dict()))
    .def("finalize", &finalize)
    .def("hasMatch", &hasMatch)
    .def("getFirstMatch", &getFirstMatch)
    .def("getMatches", &getMatches)
    .def("getEntry", &getEntry)
    .def("__len__", &nvMolKit::FilterCatalog::size)
    .add_property("pendingSize", &nvMolKit::FilterCatalog::pendingSize);
}

File: nvmolkit/filter_catalog.py
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""GPU filter catalogs for screening batches of molecules."""

from __future__ import annotations

from collections.abc import Iterable, Mapping
from dataclasses import dataclass
from enum import IntFlag

from rdkit.Chem import Mol

from nvmolkit._filterCatalog import FilterCatalog as _NativeFilterCatalog
from nvmolkit.substructure import SubstructSearchConfig

__all__ = ["FilterCatalog", "FilterCatalogEntry", "FilterCatalogPreset"]


class FilterCatalogPreset(IntFlag):
    """Filter collections bundled with RDKit."""

    PAINS_A = 1 << 1
    PAINS_B = 1 << 2
    PAINS_C = 1 << 3
    PAINS = PAINS_A | PAINS_B | PAINS_C
    BRENK = 1 << 4
    NIH = 1 << 5
    ZINC = 1 << 6
    CHEMBL_GLAXO = 1 << 7
    CHEMBL_DUNDEE = 1 << 8
    CHEMBL_BMS = 1 << 9
    CHEMBL_SURECHEMBL = 1 << 10
    CHEMBL_MLSMR = 1 << 11
    CHEMBL_INPHARMATICA = 1 << 12
    CHEMBL_LINT = 1 << 13
    CHEMBL_Glaxo = CHEMBL_GLAXO
    CHEMBL_Dundee = CHEMBL_DUNDEE
    CHEMBL_SureChEMBL = CHEMBL_SURECHEMBL
    CHEMBL_Inpharmatica = CHEMBL_INPHARMATICA
    CHEMBL = (
        CHEMBL_GLAXO
        | CHEMBL_DUNDEE
        | CHEMBL_BMS
        | CHEMBL_SURECHEMBL
        | CHEMBL_MLSMR
        | CHEMBL_INPHARMATICA
        | CHEMBL_LINT
    )
    ALL = PAINS | BRENK | NIH | ZINC | CHEMBL


def _normalize_preset(preset: FilterCatalogPreset | str | int) -> FilterCatalogPreset:
    if isinstance(preset, str):
        return FilterCatalogPreset[preset]
    return FilterCatalogPreset(preset)


@dataclass(frozen=True)
class FilterCatalogEntry:
    """Description and application metadata for one filter query."""

    id: int
    description: str
    smarts: str
    triggerCount: int
    metadata: dict[str, str]


class FilterCatalog:
    """A persistent query catalog applied to batches of target molecules.

    Query definitions remain on the CPU until :meth:`finalize` publishes them
    to the GPU. Query methods preserve target order. Match IDs refer to entries
    in insertion order and can be resolved with :meth:`getEntry`.
    """

    def __init__(
        self,
        preset: FilterCatalogPreset | str | int | None = None,
        config: SubstructSearchConfig | None = None,
    ) -> None:
        """Create an empty catalog or load one RDKit built-in filter collection."""
        if config is None:
            config = SubstructSearchConfig()
        self._native = _NativeFilterCatalog(config._as_native())
        if preset is not None:
            self._native.addPreset(int(_normalize_preset(preset)))

    def __len__(self) -> int:
        """Return the number of finalized catalog entries."""
        return len(self._native)

    @property
    def pendingSize(self) -> int:
        """Number of entries awaiting finalization."""
        return self._native.pendingSize

    def addQuery(
        self,
        query: Mol,
        description: str = "",
        triggerCount: int = 1,
        metadata: Mapping[str, str] | None = None,
    ) -> int:
        """Add an RDKit query molecule and return its stable entry ID."""
        if triggerCount <= 0:
            raise ValueError("triggerCount must be greater than zero")
        return self._native.addQuery(query, description, int(triggerCount), dict(metadata or {}))

    def addPreset(self, preset: FilterCatalogPreset | str | int) -> int:
        """Stage every entry in an RDKit built-in filter collection."""
        return self._native.addPreset(int(_normalize_preset(preset)))

    def addSmarts(
        self,
        smarts: str,
        description: str = "",
        triggerCount: int = 1,
        metadata: Mapping[str, str] | None = None,
    ) -> int:
        """Parse and add a SMARTS query, returning its stable entry ID."""
        if triggerCount <= 0:
            raise ValueError("triggerCount must be greater than zero")
        return self._native.addSmarts(smarts, description, int(triggerCount), dict(metadata or {}))

    def finalize(self) -> None:
        """Upload pending queries and publish the resulting catalog."""
        self._native.finalize()

    def getEntry(self, entryId: int) -> FilterCatalogEntry:
        """Return the definition and metadata for an entry ID."""
        entry = self._native.getEntry(int(entryId))
        return FilterCatalogEntry(
            id=entry["id"],
            description=entry["description"],
            smarts=entry["smarts"],
            triggerCount=entry["triggerCount"],
            metadata=dict(entry["metadata"]),
        )

    def hasMatch(self, molecules: Iterable[Mol]) -> list[bool]:
        """Return whether each target matches any catalog entry."""
        return self._native.hasMatch(molecules)

    def getFirstMatch(self, molecules: Iterable[Mol]) -> list[int | None]:
        """Return the first matching entry ID for each target, or ``None``."""
        return self._native.getFirstMatch(molecules)

    def getMatches(self, molecules: Iterable[Mol]) -> list[list[int]]:
        """Return all matching entry IDs for each target in catalog order."""
        return self._native.getMatches(molecules)
